### PR TITLE
feat: Reward model

### PR DESCRIPTION
**Pull request by Asher-hss**
## Description
https://github.com/camel-ai/camel/issues/889

## Motivation and Context

https://github.com/camel-ai/camel/issues/889

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/models/reward/nemetro_model.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                choices[0].logprobs.content
+                if choices and choices[0].logprobs
+                else None
+            )
+            scores = (
+                {entry.token: entry.logprob for entry in logprobs if entry}
+                if logprobs
+                else {}
+            )
+        except (KeyError, IndexError):
```
Comment (AveryYay) [inline camel/models/reward/nemetro_model.py:110]: Maybe add a log here? 
```suggestion
        except (KeyError, IndexError):
            logging.error(f"Error parsing scores: {e}")
```

File: camel/models/reward/evaluator.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            messages (List[Dict[str, str]]): A list of messages where each
+                message is a dictionary with 'role' and 'content'.
+            threshold (Dict[str, float]): A dictionary mapping score types to
+                their values.
+
+        Returns:
+            bool: A boolean indicating whether the messages pass the filter.
+        """
+        scores = self.evaluate(messages)
+        for score_type, threshold in thresholds.items():
```
Comment (AveryYay) [inline camel/models/reward/evaluator.py:59]: If the API fails to return some expected score types due to a configuration error or backend changes, would it be possible that the scores dictionary might lack certain keys?
```suggestion
        for score_type, threshold in thresholds.items():
            if score_type not in scores:
                logging.warning(f"Score type {score_type} missing in evaluation results.")
                return False
```

**Review by AveryYay — APPROVED** [review]

File: camel/models/reward/base_reward_model.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from camel.types import ModelType
+
+
+class BaseRewardModel(ABC):
+    r"""Abstract base class for reward models. Reward models are used to
+    evaluate messages and return scores based on different criteria.
+
+    Subclasses should implement the 'evaluate' and 'get_scores_types' methods.
+    """
+
```
Comment (Wendong-Fan) [inline camel/models/reward/base_reward_model.py:27]: mock review

File: camel/models/reward/base_reward_model.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        criteria.
+
+        Args:
+            messages (List[Dict[str, str]]): A list of messages where each
+                message is a dictionary with 'role' and 'content'.
+
+        Returns:
+            Dict[str, float]: A dictionary mapping score types to their values.
+        """
+        pass
```
Comment (koch3092) [inline camel/models/reward/base_reward_model.py:50]: It is better to throw a `NotImplementedError` than these `pass`

**Review by koch3092 — COMMENTED** [review]
thanks @Asher-hss , left some comments below.

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @Asher-hss , as mentioned in the issue, we also want to support ` skywork (Seq. Classifier)`, could you also add this model?

**Comment by Asher-hss** [comment]
> Thanks @Asher-hss , as mentioned in the issue, we also want to support ` skywork (Seq. Classifier)`, could you also add this model?

Thanks wendong，I plan to update this model today.

File: camel/models/reward/evaluator.py (hunk 2 above)
Comment (Asher-hss) [reply camel/models/reward/evaluator.py:59]: good suggestion！

File: camel/models/reward/base_reward_model.py (hunk 4 above)
Comment (Asher-hss) [reply camel/models/reward/base_reward_model.py:50]: Follow camel basemodel
@Wendong-Fan

File: camel/models/reward/nemetro_model.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        api_key = api_key or os.environ.get("NVIDIA_API_KEY")
+        super().__init__(model_type, api_key, url)
+        self._client = OpenAI(
+            timeout=60,
+            max_retries=3,
+            base_url=self.url,
+            api_key=self.api_key,
+        )
+
+    @api_keys_required("NVIDIA_API_KEY")
```
[the remark this answers is not in the document]
Comment (Asher-hss) [reply camel/models/reward/nemetro_model.py:57]: follow other model's format

File: camel/models/reward/base_reward_model.py (hunk 4 above)
Comment (Wendong-Fan) [reply camel/models/reward/base_reward_model.py:50]: Thanks @koch3092 , I think normally using `pass` is ok

File: camel/models/reward/nemetro_model.py (hunk 1 above)
Comment (Wendong-Fan) [reply camel/models/reward/nemetro_model.py:110]: Yeah, I think we shouldn't add the error msg into score dict, better raise error

File: camel/models/reward/nemetro_model.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    def evaluate(self, messages: List[OpenAIMessage]) -> Dict[str, float]:
```
Comment (Wendong-Fan) [inline camel/models/reward/nemetro_model.py:58]: set messages type as List[Dict[str, str]] would suit for more general use case

File: camel/models/reward/nemetro_model.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        r"""Evaluate the messages using the Nemetro model.
+
+        Args:
+            messages (List[OpenAIMessage]): A list of messages where each
+                message is an OpenAIMessage object.
+
+        Returns:
+            ChatCompletion: A ChatCompletion object with the scores.
```
Comment (Wendong-Fan) [inline camel/models/reward/nemetro_model.py:66]: not align with type hint

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @Asher-hss ! Left some comments below and added one more commit here based on the review, feel free to review the change, I think we can add the skywork model in another PR to let this PR merged first, but please add **unit test** first before the merge, thanks!

**Review by Wendong-Fan — APPROVED** [review]

**Comment by lightaime** [comment]
The model is called `[nemotron](https://build.nvidia.com/nvidia/nemotron-4-340b-instruct)` not `nemetro`can you fix that asap?

**Comment by Asher-hss** [comment]
> The model is called `[nemotron](https://build.nvidia.com/nvidia/nemotron-4-340b-instruct)` not `nemetro`can you fix that asap?

I create a new pr to fix it now.

**Pull request merged**